// mimir/index.js
(function () {

  function tokenize(text) {
    return text
      .replace(/á/gi, 'a')
      .replace(/é/gi, 'e')
      .replace(/í/gi, 'i')
      .replace(/ó/gi, 'o')
      .replace(/ú/gi, 'u')
      .replace(/ñ/gi, 'n')
      .replace(/ el | la | los | que | las | a | muy | de | tus | con | mucho | poco | muchos | pocos | algo | algunos /gi, ' ')
      .replace(/El |La |Los |Que |Las |A |Muy |De |Tus |Con mucho |Poco |Puchos |Pocos |Algo |Algunos /g, ' ')
      // .replace(/ alguna | algunas | alguno | algunos | ambos | ampleamos | ante | antes | aquel | aquellas | aquellos | aqui | arriba | atras | bajo | bastante | cada | cierta | ciertas | cierto | ciertos | como | con | conseguimos | conseguir | consigo | consigue | consiguen | consigues | cual | cuando | dentro | desde | donde | dos | el | ellas | ellos | emplean | emplear | empleas | empleo | en | encima | entonces | entre | era | eramos | eran | eras | eres | es | esta | estaba | estado | estamos | estan | estoy | fin | fue | fueron | fui | fuimos | ha | hace | hacemos | hacen | hacer | haces | hago | incluso | intenta | intentamos | intentan | intentar | intentas | intento | ir | la | largo | las | lo | los | mientras | mio | modo | muchos | muy | nos | nosotros | otro | para | pero | podemos | poder | podria | podriamos | podrian | podrias | por | por que | porque | primero | puede | pueden | puedo | quien | sabe | sabemos | saben | saber | sabes | ser | si | siendo | sin | sobre | solamente | solo | somos | soy | su | sus | tambien | teneis | tenemos | tener | tengo | tiempo | tiene | tienen | todo | trabaja | trabajamos | trabajan | trabajar | trabajas | trabajo | tras | tuyo | ultimo | un | una | unas | uno | unos | usa | usamos | usan | usar | usas | uso | va | valor | vamos | van | vaya | verdad | verdadera | verdadero | vosotras | vosotros | voy | yo | ahora | al | algo | alrededor | anterior | apenas | aproximadamente | aunque | ayer | buen | buena | buenas | bueno | buenos | casi | cerca | cinco | conocer | considera | contra | cosas | creo | cuales | cualquier | cuanto | cuatro | cuenta | da | dado | dan | dar | de | debe | deben | debido | decir | del | dice | dicen | dicho | dieron | diferente | diferentes | dijeron | dijo | dio | durante | e | ejemplo | ella | ello | embargo | encuentra | esa | esas | ese | eso | esos | estaban | estar | estara | estan | estas | este | esto | estos | estuvo | ex | existe | existen | explican | expresan | fuera | gran | grandes | habia | habian | haber | habra | hacerlo | hacia | haciendo | han | hasta | hay | haya | he | hecho | hemos | hicieron | hizo | hoy | hubo | igual | indica | informa | junto | lado | le | les | llega | lleva | llevar | luego | lugar | mas | manera | mayor | me | mediante | mejor | menciona | menos | mi | misma | mismas | mismo | mismos | momento | mucha | muchas | mucho | nada | nadie | ni | ningun | ninguna | ningunas | ninguno | ningunos | no | nosotras | nuestra | nuestras | nuestro | nuestros | nueva | nuevas | nuevo | nuevos | nunca | o | ocho | otra | otras | otros | parece | parte | partir | pasada | pasado | pesar | poca | pocas | poco | pocos | podra | podran | podria | podrían | poner | posible | proximo | proximos | primer | primera | primeros | principalmente | propia | propias | propio | propios | pudo | pueda | pues | que | queda | queremos | quien | quienes | quiere | realiza | realizado | realizar | respecto | si | se | sea | sean | segun | segunda | segundo | seis | sido | siempre | siete | sigue | siguiente | sino | sola | solas | solos | son | tal | tampoco | tan | tanto | tenia | tendra | tendran | tenga | tenido | tercera | toda | todas | todavía | todos | total | trata | tres | tuvo | usted | varias | varios | veces | ver | vez | y | ya /g, ' ')
      // .replace(/Alguna |Algunas |Alguno |Algunos |Ambos |Ampleamos |Ante |Antes |Aquel |Aquellas |Aquellos |Aqui |Arriba |Atras |Bajo |Bastante |Cada |Cierta |Ciertas |Cierto |Ciertos |Como |Con |Conseguimos |Conseguir |Consigo |Consigue |Consiguen |Consigues |Cual |Cuando |Dentro |Desde |Donde |Dos |El |Ellas |Ellos |Emplean |Emplear |Empleas |Empleo |En |Encima |Entonces |Entre |Era |Eramos |Eran |Eras |Eres |Es |Esta |Estaba |Estado |Estamos |Estan |Estoy |Fin |Fue |Fueron |Fui |Fuimos |Ha |Hace |Hacemos |Hacen |Hacer |Haces |Hago |Incluso |Intenta |Intentamos |Intentan |Intentar |Intentas |Intento |Ir |La |Largo |Las |Lo |Los |Mientras |Mio |Modo |Muchos |Muy |Nos |Nosotros |Otro |Para |Pero |Podemos |Poder |Podria |Podriamos |Podrian |Podrias |Por |Por que |Porque |Primero |Puede |Pueden |Puedo |Quien |Sabe |Sabemos |Saben |Saber |Sabes |Ser |Si |Siendo |Sin |Sobre |Solamente |Solo |Somos |Soy |Su |Sus |Tambien |Teneis |Tenemos |Tener |Tengo |Tiempo |Tiene |Tienen |Todo |Trabaja |Trabajamos |Trabajan |Trabajar |Trabajas |Trabajo |Tras |Tuyo |Ultimo |Un |Una |Unas |Uno |Unos |Usa |Usamos |Usan |Usar |Usas |Uso |Va |Valor |Vamos |Van |Vaya |Verdad |Verdadera |Verdadero |Vosotras |Vosotros |Voy |Yo |Ahora |Al |Algo |Alrededor |Anterior |Apenas |Aproximadamente |Aunque |Ayer |Buen |Buena |Buenas |Bueno |Buenos |Casi |Cerca |Cinco |Conocer |Considera |Contra |Cosas |Creo |Cuales |Cualquier |Cuanto |Cuatro |Cuenta |Da |Dado |Dan |Dar |De |Debe |Deben |Debido |Decir |Del |Dice |Dicen |Dicho |Dieron |Diferente |Diferentes |Dijeron |Dijo |Dio |Durante |E |Ejemplo |Ella |Ello |Embargo |Encuentra |Esa |Esas |Ese |Eso |Esos |Estaban |Estar |Estara |Estan |Estas |Este |Esto |Estos |Estuvo |Ex |Existe |Existen |Explican |Expresan |Fuera |Fran |Grandes |Habia |Habian |Haber |Habra |Hacerlo |Hacia |Haciendo |Han |Hasta |Hay |Haya |He |Hecho |Hemos |Hicieron |Hizo |Hoy |Hubo |Igual |Indica |Informa |Junto |Lado |Le |Les |Llega |Lleva |Llevar |Luego |Lugar |Mas |Manera |Mayor |Me |Mediante |Mejor |Menciona |Menos |Mi |Misma |Mismas |Mismo |Mismos |Momento |Mucha |Muchas |Mucho |Nada |Nadie |Ni |Ningun |Ninguna |Ningunas |Ninguno |Ningunos |No |Nosotras |Nuestra |Nuestras |Nuestro |Nuestros |Nueva |Nuevas |Nuevo |Nuevos |Nunca |O |Ocho |Otra |Otras |Otros |Parece |Parte |Partir |Pasada |Pasado |Pesar |Poca |Pocas |Poco |Pocos |Podra |Podran |Podria |Podrían |Poner |Posible |Proximo |Proximos |Primer |Primera |Primeros |Principalmente |Propia |Propias |Propio |Propios |Pudo |Pueda |Pues |Que |Queda |Queremos |Quien |Quienes |Quiere |Realiza |Realizado |Realizar |Respecto |Si |Se |Sea |Sean |Segun |Segunda |Segundo |Seis |Sido |Siempre |Siete |Sigue |Siguiente |Sino |Sola |Solas |Solos |Son |Tal |Tampoco |Tan |Tanto |Tenia |Tendra |Tendran |Tenga |Tenido |Tercera |Toda |Todas |Todavía |Todos |Total |Trata |Tres |Tuvo |Usted |Varias |Varios |Veces |Ver |Vez |Y |Ya  /g, ' ')
      .replace(/\W/g, ' ')
      .replace(/\s\s+/g, ' ')
      .split(' ').map(function (s) {
        return s.toLowerCase();
      });
  }

  function extractDictionary(textArray) {
    var dict = {},
      keys = [],
      words;
    textArray = Array.isArray(textArray) ? textArray : [textArray];
    textArray.forEach(function (text) {
      words = tokenize(text);
      words.forEach(function (word) {
        word = word.toLowerCase();
        if (!dict[word] && word !== '') {
          dict[word] = 1;
          keys.push(word);
        } else {
          dict[word] += 1;
        }
      });
    });

    return {
      words: keys,
      dict: dict
    };
  }

  function bow(text, vocabulary) {
    var dict = extractDictionary([text]).dict,
      vector = [];

    vocabulary.words.forEach(function (word) {
      vector.push(dict[word] || 0);
    });
    return vector;
  }

  function tf(word, text) {
    var input = word.toLowerCase();
    var dict = extractDictionary(text).dict;
    return dict[input] / tokenize(text).length;
  }

  function wordInDocsCount(word, textlist) {
    var sum = 0;
    textlist.forEach(function (text) {
      sum += tokenize(text).indexOf(word) > -1 ? 1 : 0;
    });
    return sum;
  }

  function idf(word, textlist) {
    return Math.log(textlist.length / (1 + wordInDocsCount(word, textlist)));
  }

  function tfidf(word, text, textlist) {
    return tf(word, text) * idf(word, textlist);
  }

  module.exports = {
    dict: extractDictionary,
    bow: bow,
    tfidf: tfidf,
    tokenize: tokenize
  };

}());
